### PR TITLE
Add run_and_handle_error to check for expected errors

### PR DESCRIPTION
**Pull request by rossf7**
Towards giantswarm/giantswarm#18384

Adds a new `run_and_handle_error` function that checks if an error occurs and returns exit code 0.
This is to avoid complicating the existing `run_and_log` function.

It's needed because `go test` returns an error if no go files match the build tag specified.

```
go test -v -tags=functional ./integration/test/metrics
package github.com/giantswarm/app-test-suite/integration/test/metrics: build constraints exclude all Go files in /Users/rossf7/src/giantswarm/app-test-suite/integration/test/metrics
```

### CODE REVIEW

File: step_exec_lib/utils/processes.py [hunk 1]
```diff
@@ -10,6 +10,29 @@ def run_and_log(args: List[str], **kwargs: Any) -> subprocess.CompletedProcess:
     logger.info(" ".join(args))
     if "text" not in kwargs:
         kwargs["text"] = True
+
     run_res = subprocess.run(args, **kwargs)  # nosec
     logger.info(f"Command executed, exit code: {run_res.returncode}.")
     return run_res
+
+
+def run_and_handle_error(args: List[str], error_text: str, **kwargs: Any) -> int:
```
Comment (piontec) [inline step_exec_lib/utils/processes.py:19]:
```suggestion
def run_and_handle_error(args: List[str], epected_error_text: str, **kwargs: Any) -> int:
```

File: step_exec_lib/utils/processes.py [hunk 2]
```diff
@@ -10,6 +10,29 @@ def run_and_log(args: List[str], **kwargs: Any) -> subprocess.CompletedProcess:
...
+
+    except subprocess.CalledProcessError as e:
+        logger.info(f"CalledProcessError: {e}.")
+
+        if error_text in e.stderr:
+            logger.info(f"Found expected error text '{error_text}', exit code: 0")
+            return 0
+        else:
+            logger.info(f"Did not find error text '{error_text}', exit code: 1")
+            return 1
```
Comment (piontec) [inline step_exec_lib/utils/processes.py:38]: Is it OK to return `run_res.returncode` to not lose the original error code?

**Review by piontec — APPROVED** [review]
LGTM, with some minor comments

File: step_exec_lib/utils/processes.py [hunk 3]
```diff
@@ -10,6 +10,29 @@ def run_and_log(args: List[str], **kwargs: Any) -> subprocess.CompletedProcess:
...
     run_res = subprocess.run(args, **kwargs)  # nosec
     logger.info(f"Command executed, exit code: {run_res.returncode}.")
     return run_res
+
+
+def run_and_handle_error(args: List[str], error_text: str, **kwargs: Any) -> int:
+    logger.info("Running command:")
+    logger.info(" ".join(args))
+    if "text" not in kwargs:
+        kwargs["text"] = True
```
Comment (MarcelMue) [inline step_exec_lib/utils/processes.py:23]: Maybe I don't know enough about python but why is this needed?

Comment (rossf7) [reply step_exec_lib/utils/processes.py:23]: @MarcelMue This is the same as `run_and_log` just with the extra error handling needed for the go test executor.

@piontec Can you help with why we need this?

File: step_exec_lib/utils/processes.py (hunk 2 above)
Comment (rossf7) [reply step_exec_lib/utils/processes.py]: Yes sure. I changed this and the param name.

File: step_exec_lib/utils/processes.py (hunk 3 above)
Comment (piontec) [reply step_exec_lib/utils/processes.py:23]: It's part of how `run` works, see: https://docs.python.org/3/library/subprocess.html#subprocess.run

Basically it means "parse stdout as encoded text".

**Review by MarcelMue — APPROVED** [review]
Thanks for the explanation - works for me :+1:
